File: app/src/main/java/com/maxandron/daggerissue325java/foo/FooActivity.java
package com.maxandron.daggerissue325java.foo;

import android.os.Bundle;
import android.support.annotation.Nullable;

import com.maxandron.daggerissue325java.BaseActivity;
import com.maxandron.daggerissue325java.MyApplication;

import javax.inject.Inject;

public class FooActivity extends BaseActivity {

    @Inject
    Foo foo;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    protected void setupComponent() {
        MyApplication.get(this).getAppComponent().plus(new FooProvider(this)).inject(this);
    }
}

File: app/src/main/java/com/maxandron/daggerissue325java/foo/FooActivity.kt
package com.maxandron.daggerissue325java.foo

import android.os.Bundle

import com.maxandron.daggerissue325java.BaseActivity
import com.maxandron.daggerissue325java.MyApplication

import javax.inject.Inject

class FooActivity : BaseActivity() {

    @Inject
    lateinit var foo: Foo

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    override fun setupComponent() {
        MyApplication[this].appComponent.plus(FooProvider(this)).inject(this)
    }
}
